traffic_ops/testing/api/v4/cachegroups_parameters_test.go

```diff
@@ -155,3 +274,201 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
+	}
+
+	res := sort.SliceIsSorted(sortedList, func(p, q int) bool {
+		return sortedList[p] < sortedList[q]
+	})
+	if !res {
+		t.Errorf("list is not sorted by their ID: %v", sortedList)
+	}
+}
+
+func SortTestCacheGroupParametersDesc(t *testing.T) {
+	resp, _, err := TOSession.GetAllCacheGroupParameters(client.RequestOptions{})
+	if err != nil {
+		t.Errorf("Expected no error, but got error in CacheGroup Parameters with default ordering: %v - alerts: %+v", err, resp.Alerts)
+	}
+	cachegroupParameters := resp.Response
+	respAsc := cachegroupParameters.CacheGroupParameters
+	if len(respAsc) < 1 {
+		t.Fatal("Need at least one CacheGroup Parameters in Traffic Ops to test CacheGroup Parameters sort ordering")
+	}
+
+	opts := client.NewRequestOptions()
+	opts.QueryParameters.Set("sortOrder", "desc")
+	resp, _, err = TOSession.GetAllCacheGroupParameters(opts)
+	if err != nil {
+		t.Errorf("Expected no error, but got error in CacheGroup Parameters with Descending ordering: %v - alerts: %+v", err, resp.Alerts)
+	}
+	cachegroupParameters = resp.Response
+	respDesc := cachegroupParameters.CacheGroupParameters
+	if len(respDesc) < 1 {
+		t.Fatal("Need at least one CacheGroup Parameters in Traffic Ops to test CacheGroup Parameters sort ordering")
+	}
+
+	if len(respAsc) != len(respDesc) {
+		t.Fatalf("Traffic Ops returned %d CacheGroup Parameters using default sort order, but %d CacheGroup Parameters when sort order was explicitly set to descending", len(respAsc), len(respDesc))
+	}
+
+	// reverse the descending-sorted response and compare it to the ascending-sorted one
+	// TODO ensure at least two in each slice? A list of length one is
+	// trivially sorted both ascending and descending.
+	for start, end := 0, len(respDesc)-1; start < end; start, end = start+1, end-1 {
+		respDesc[start], respDesc[end] = respDesc[end], respDesc[start]
+	}
+	if *respDesc[0].Parameter != *respAsc[0].Parameter {
+		t.Errorf("CacheGroup Parameters responses are not equal after reversal: Asc: %v - Desc: %v", *respDesc[0].Parameter, *respAsc[0].Parameter)
+	}
+}
+
+func CreateTestCacheGroupParametersWithInvalidData(t *testing.T) {
+	resp, _, err := TOSession.GetParameters(client.RequestOptions{})
+	parametersResp := resp.Response[0]
+	parameterID := parametersResp.ID
+	//Create CacheGroupParameters with Invalid Cachegroup ID
+	cgparameters, reqInf, err := TOSession.CreateCacheGroupParameter(0, parameterID, client.RequestOptions{})
+	if reqInf.StatusCode != http.StatusNotFound {
+		t.Errorf("Expected 404 status code, got %v", reqInf.StatusCode)
+	}
+	if err == nil {
+		t.Errorf("Expected cachegroup not found, but found Alerts %v", cgparameters.Alerts)
+	}
+
+	cgresp, _, err := TOSession.GetCacheGroups(client.RequestOptions{})
+	cachegroupResp := cgresp.Response[0]
+	cachegroupID := cachegroupResp.ID
+
+	//Create CacheGroupParameters with Invalid Parameter ID
+	cgparametersresp, reqInf, err := TOSession.CreateCacheGroupParameter(*cachegroupID, 0, client.RequestOptions{})
```

Comment:
This will segfault if the Cache Group ID was null or undefined in the response

Reply:
added null checks